### PR TITLE
use cmdline arguments for specifying topics

### PR DESCRIPTION
**Pull request by Karsten1987**
connects to ros2/ros2#327

### CODE REVIEW

File: demo_nodes_cpp/package.xml [hunk 1]
```diff
@@ -14,12 +14,14 @@
   <buildtool_depend>rosidl_default_generators</buildtool_depend>
 
   <build_depend>example_interfaces</build_depend>
+  <build_depend>c_utilities</build_depend>
```
Comment (mikaelarguedas) [inline demo_nodes_cpp/package.xml]: Nit: alphabetical order, same below

File: demo_nodes_cpp/src/services/add_two_ints_client.cpp [hunk 2]
```diff
@@ -16,12 +16,22 @@
 #include <iostream>
 #include <memory>
 
+#include "c_utilities/cmdline_parser.h"
 #include "rclcpp/rclcpp.hpp"
 
 #include "example_interfaces/srv/add_two_ints.hpp"
```
Comment (mikaelarguedas) [inline demo_nodes_cpp/src/services/add_two_ints_client.cpp]: can this be moved before rclcpp.h as well ?

File: demo_nodes_cpp/src/services/add_two_ints_client.cpp [hunk 3]
```diff
@@ -16,12 +16,22 @@
...
 
 using namespace std::chrono_literals;
 
+void print_usage()
+{
+  printf("Usage for add_two_ints_client app:\n");
+  printf("add_two_ints_client [-t topic_name] [-h]\n");
+  printf("options:\n");
+  printf("-h : Print this help function.\n");
+  printf("-t topic_name : Specify the topic on which to publish. Defaults to add_two_ints.\n");
```
Comment (mikaelarguedas) [inline demo_nodes_cpp/src/services/add_two_ints_client.cpp]: Nit: maybe request topic would be clearer here

File: demo_nodes_cpp/src/services/add_two_ints_client.cpp [hunk 4]
```diff
@@ -45,7 +55,17 @@ int main(int argc, char ** argv)
 
   auto node = rclcpp::Node::make_shared("add_two_ints_client");
 
-  auto client = node->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");
+  if (cli_option_exist(argv, argv+argc, "-h")) {
```
Comment (mikaelarguedas) [inline demo_nodes_cpp/src/services/add_two_ints_client.cpp]: Nit: whitespace before and after + operator, same verywhere in this PR

File: demo_nodes_cpp/src/services/add_two_ints_server.cpp [hunk 5]
```diff
@@ -15,10 +15,20 @@
...
 
 #include "example_interfaces/srv/add_two_ints.hpp"
 
+void print_usage()
+{
+  printf("Usage for add_two_ints_server app:\n");
+  printf("add_two_ints_server [-t topic_name] [-h]\n");
+  printf("options:\n");
+  printf("-h : Print this help function.\n");
+  printf("-t topic_name : Specify the topic on which to publish. Defaults to add_two_ints.\n");
```
Comment (mikaelarguedas) [inline demo_nodes_cpp/src/services/add_two_ints_server.cpp]: maybe Response (or Reply?) topic would be clearer here

File: demo_nodes_cpp/src/topics/listener.cpp [hunk 6]
```diff
@@ -15,10 +15,20 @@
...
 
 #include "std_msgs/msg/string.hpp"
 
+void print_usage()
+{
+  printf("Usage for listener app:\n");
+  printf("listener [-t topic_name] [-h]\n");
+  printf("options:\n");
+  printf("-h : Print this help function.\n");
+  printf("-t topic_name : Specify the topic on which to publish. Defaults to chatter.\n");
```
Comment (mikaelarguedas) [inline demo_nodes_cpp/src/topics/listener.cpp]: Do you mean topic to subscribe to ?

**Comment by mikaelarguedas** [comment]
@Karsten1987 oops just saw this is still "in progress", sorry if it was not ready for review

File: demo_nodes_cpp/src/services/add_two_ints_client.cpp (hunk 4 above)
Comment (wjwwood) [inline demo_nodes_cpp/src/services/add_two_ints_client.cpp]: If these are C functions, `cli_option_exist()` and others, then I think you ought to namespace them, e.g. `utilities_cli_option_exist()`.

File: demo_nodes_cpp/src/services/add_two_ints_client.cpp [hunk 7]
```diff
@@ -45,7 +55,17 @@ int main(int argc, char ** argv)
...
+    print_usage();
+    return 0;
+  }
+
+  auto topic = cli_get_option(argv, argv+argc, "-t");
+  if (!topic) {
+    topic = const_cast<char *>("add_two_ints");
```
Comment (wjwwood) [inline demo_nodes_cpp/src/services/add_two_ints_client.cpp]: This seems dubious. What if the input to `cli_get_option` is really a string literal...

File: demo_nodes_cpp/src/services/add_two_ints_client.cpp (hunk 3 above)
Comment (wjwwood) [inline demo_nodes_cpp/src/services/add_two_ints_client.cpp]: Wouldn't just "service name" be clearer?

File: demo_nodes_cpp/src/services/add_two_ints_server.cpp (hunk 5 above)
Comment (wjwwood) [inline demo_nodes_cpp/src/services/add_two_ints_server.cpp]: Just service I think.

File: demo_nodes_cpp/src/services/add_two_ints_client.cpp [hunk 8]
```diff
@@ -45,7 +55,17 @@ int main(int argc, char ** argv)
...
 
-  auto client = node->create_client<example_interfaces::srv::AddTwoInts>("add_two_ints");
+  if (cli_option_exist(argv, argv + argc, "-h")) {
+    print_usage();
+    return 0;
+  }
+
+  auto topic = cli_get_option(argv, argv + argc, "-t");
+  if (!topic) {
+    topic = const_cast<char *>("add_two_ints");
```
Comment (wjwwood) [inline demo_nodes_cpp/src/services/add_two_ints_client.cpp]: 🚨 bad bad bad 🚨

File: demo_nodes_cpp/src/services/add_two_ints_server.cpp [hunk 9]
```diff
@@ -36,8 +46,18 @@ int main(int argc, char ** argv)
...
   auto node = rclcpp::Node::make_shared("add_two_ints_server");
 
+  if (cli_option_exist(argv, argv + argc, "-h")) {
+    print_usage();
+    return 0;
+  }
+
+  auto topic = cli_get_option(argv, argv + argc, "-t");
+  if (!topic) {
+    topic = const_cast<char *>("add_two_ints");
```
Comment (wjwwood) [inline demo_nodes_cpp/src/services/add_two_ints_server.cpp]: 🚨 bad bad bad 🚨

File: demo_nodes_cpp/src/topics/listener.cpp (hunk 6 above)
Comment (wjwwood) [inline demo_nodes_cpp/src/topics/listener.cpp]: 👍

File: demo_nodes_cpp/src/topics/listener.cpp [hunk 10]
```diff
@@ -29,8 +39,17 @@ int main(int argc, char * argv[])
...
   auto node = rclcpp::Node::make_shared("listener");
 
+  if (cli_option_exist(argv, argv + argc, "-h")) {
+    print_usage();
+    return 0;
+  }
+
+  auto topic = cli_get_option(argv, argv + argc, "-t");
+  if (!topic) {
+    topic = const_cast<char *>("chatter");
```
Comment (wjwwood) [inline demo_nodes_cpp/src/topics/listener.cpp]: 🚨 bad bad bad 🚨

File: demo_nodes_cpp/src/topics/talker.cpp [hunk 11]
```diff
@@ -28,7 +38,16 @@ int main(int argc, char * argv[])
...
 
-  auto chatter_pub = node->create_publisher<std_msgs::msg::String>("chatter", custom_qos_profile);
+  if (cli_option_exist(argv, argv + argc, "-h")) {
+    print_usage();
+    return 0;
+  }
+
+  auto topic = cli_get_option(argv, argv + argc, "-t");
+  if (!topic) {
+    topic = const_cast<char *>("chatter");
```
Comment (wjwwood) [inline demo_nodes_cpp/src/topics/talker.cpp]: 🤖 danger, danger will robinson 🤖

**Pull request merged**